### DISCUSSION TITLE
Accelerated operations on m1 mac using blas

### DISCUSSION THREAD

**Discussion by jimmiebtlr** [rust-ndarray/ndarray#1265] [Q&A] [answered]
I'm having trouble finding the right combination to get ndarray + blas acceleration working on mac (or how I'm checking is incorrect).

Currently in dependencies of Cargo.toml I have the following lines.
```
ndarray = {version="0.15.6", features=["blas"]}
blas-src = { version = "0.8", features = ["accelerate"] }
```

My program is running alot of multiplications and additions of a 1d array with 50k elements, but gpu according to activity monitor never goes up, and speeds seem similar.

I've also tried
```
ndarray = { version = "0.15.0", features = ["blas"] }
blas-src = { version = "0.8", features = ["openblas"] }
openblas-src = { version = "0.10", features = ["cblas", "system"] }
```

**Comment by jturner314** [comment, marked as answer]
The [main Apple documentation page about Accelerate](https://developer.apple.com/accelerate/) seems to suggest that it's CPU-only:

> The Accelerate framework provides high-performance, energy-efficient computation on the CPU by leveraging its vector-processing capability.
>
> ...
>
> The Accelerate framework provides BLAS and LAPACK libraries for performing linear algebra on dense vectors and matrices. Accelerate’s BLAS and LAPACK implementations abstract the processing capability of the CPU so code written for them will execute the appropriate instructions for the processor available at runtime. This means that both BLAS and LAPACK are optimized for high performance and low-energy consumption.

[OpenBLAS is also CPU-only.](https://github.com/xianyi/OpenBLAS/wiki/faq#recent_hardware) I think all of the options provided by [`blas-src`](https://crates.io/crates/blas-src) are CPU-only.

Also note that the `blas` feature only affects matrix multiplications.

Most matrix and linear algebra crates are CPU-only, but there are libraries which use the GPU. I'm not sure what options are available for the M1. [`arrayfire`](https://crates.io/crates/arrayfire) with the OpenCL backend may be suitable.

**Reply by jimmiebtlr** [reply]
That's what I was missing, thank you.

**Reply by antimora** [reply]
@jimmiebtlr You can also try tch-rs [(bindings for Torch) that supports CUDA cards](https://github.com/LaurentMazare/tch-rs). The latest code (on main) supports downloading torch lib automatically for Mac m1. We use for our Burn backend: https://github.com/burn-rs/burn
